### DISCUSSION TITLE
Priority of map rules?

### DISCUSSION THREAD

**Discussion by trevorgithub** [mitmproxy/mitmproxy#7189] [Q&A]
I understand that I can use mitmproxy to map requests to remotes or locals using map-remote and map-local.  From the [docs](https://docs.mitmproxy.org/stable/overview-features/#map-local) I see that an arbitrary number of patterns can be specified.

I'm wondering if I have two rules that can both apply to the same request, which takes precedence if there is some overlap? Is it the one that is more specific?  Is it related to the order in which the mapping rules are applied?

For example, imagine I have URLs like this:
```
http://example.com/path/code.js
http://example.com/path/api/fetch-stuff
```

If I have rules like this:
```
mitmproxy --map-remote "|http://example.com/path|http://localhost:8081" --map-remote "|http://example.com/path/api|http://example.com/path/api"
```

(Please forgive me, if the syntax is wrong.  Very new to mitmproxy :))

Essentially, I want API requests to pass through to the original server, but requests for static assets (anything not API related) to be redirected to my local server.  The requests coming in share the same host and starting path; where they diverge is that API requests include API in the path, while other requests for static assets does not.

**Comment by Prinzhorn** [comment]
Code is there https://github.com/mitmproxy/mitmproxy/blob/2c1802692ddb99b8d02a353d8382772029168616/mitmproxy/addons/mapremote.py#L58-L68

I think we should make this more clear or even configurable. Right now your first rule will change the host to "localhost" which will then cause the second rule to no longer match the "example.com" pattern. But in general all rules are applied in order as long as they match.

> Essentially, I want API requests to pass through to the original server, but requests for static assets (anything not API related) to be redirected to my local server. The requests coming in share the same host and starting path; where they diverge is that API requests include API in the path, while other requests for static assets does not.

In this case I'd use a single rule with a negation on "/api". E.g. using [filters](https://docs.mitmproxy.org/stable/concepts-filters/) and something like `!(~u "http://example.com/path/api/.*")`

```
|!(~u "http://example.com/path/api/.*")|http://example.com/.*|http://localhost:8081
```

(honestly not sure about the syntax, have never used this feature)

**Reply by trevorgithub** [reply]
Thank you!  I'll test this out as soon as I can, and will mark the answer appropriately.
